File: MacOS Tahoe 26/ESP32_PN532_SmartCardReader/ESP32_PN532_SmartCardReader/NDEFEncoder.swift
//********************************************************************************************************//
// Nome Modulo: NDEFEncoder.swift
// Descrizione: Generatore e formattatore avanzato di flussi binari conformi allo standard NFC Forum NDEF.
//              Supporta la concatenazione dinamica di record multipli (Testo e URI).
//********************************************************************************************************//

import Foundation

enum NDEFType: String, CaseIterable, Identifiable {
    case url = "URI/URL"
    case text = "Testo"
    var id: String { self.rawValue }
}

struct NDEFWriteRecord: Identifiable {
    let id = UUID()
    var type: NDEFType = .url
    var value: String = ""
}

class NDEFEncoder {
    
    /**
     * Prende un array di record configurati dall'utente e genera il blocco binario TLV allineato a 4 byte
     */
    static func encodeMultipleRecords(records: [NDEFWriteRecord]) -> [UInt8] {
        var ndefMessageBytes: [UInt8] = []
        let validRecords = records.filter { !$0.value.isEmpty }
        
        guard !validRecords.isEmpty else { return [] }
        
        for (index, record) in validRecords.enumerated() {
            let isFirst = (index == 0)
            let isLast = (index == validRecords.count - 1)
            
            // Calcolo dei flag di controllo NDEF (MB, ME, CF=0, SR=1, IL=0, TNF=0x01 Well-Known)
            var headerByte: UInt8 = 0x01 // TNF predefinito
            if isFirst { headerByte |= 0x80 } // MB = 1
            if isLast  { headerByte |= 0x40 } // ME = 1
            headerByte |= 0x10                // SR = 1 (Short Record costantemente a 1 byte di lunghezza)
            
            switch record.type {
            case .url:
                if let uriBytes = compileURIRecord(header: headerByte, urlString: record.value) {
                    ndefMessageBytes.append(contentsOf: uriBytes)
                }
            case .text:
                if let textBytes = compileTextRecord(header: headerByte, textString: record.value) {
                    ndefMessageBytes.append(contentsOf: textBytes)
                }
            }
        }
        
        // Incapsulamento nel blocco di trasmissione TLV (Type-Length-Value) per la memoria dell'NTAG
        var tlvPayload: [UInt8] = []
        tlvPayload.append(0x03) // T: NDEF Message Type
        tlvPayload.append(UInt8(ndefMessageBytes.count & 0xFF)) // L: Lunghezza totale del messaggio NDEF
        tlvPayload.append(contentsOf: ndefMessageBytes) // V: Array di record concatenati
        tlvPayload.append(0xFE) // Terminatore del blocco TLV
        
        // Allineamento geometrico a pagine di 4 byte (Padding con 0x00)
        let paddingNeeded = (4 - (tlvPayload.count % 4)) % 4
        if paddingNeeded > 0 {
            tlvPayload.append(contentsOf: [UInt8](repeating: 0x00, count: paddingNeeded))
        }
        
        return tlvPayload
    }
    
    private static func compileURIRecord(header: UInt8, urlString: String) -> [UInt8]? {
        var uriBody = urlString
        var prefixCode: UInt8 = 0x00
        
        if urlString.hasPrefix("https://www.") { prefixCode = 0x02; uriBody = String(urlString.dropFirst(12)) }
        else if urlString.hasPrefix("http://www.") { prefixCode = 0x01; uriBody = String(urlString.dropFirst(11)) }
        else if urlString.hasPrefix("https://") { prefixCode = 0x04; uriBody = String(urlString.dropFirst(8)) }
        else if urlString.hasPrefix("http://") { prefixCode = 0x03; uriBody = String(urlString.dropFirst(7)) }
        else if urlString.hasPrefix("mailto:") { prefixCode = 0x06; uriBody = String(urlString.dropFirst(7)) }
        
        let bodyBytes = Array(uriBody.utf8)
        let payloadLength = bodyBytes.count + 1
        
        var record: [UInt8] = []
        record.append(header)
        record.append(0x01) // Type Length ("U")
        record.append(UInt8(payloadLength & 0xFF))
        record.append(0x55) // Type: "U"
        record.append(prefixCode)
        record.append(contentsOf: bodyBytes)
        return record
    }
    
    private static func compileTextRecord(header: UInt8, textString: String) -> [UInt8]? {
        let textBytes = Array(textString.utf8)
        let langBytes = Array("en".utf8) // Codice lingua predefinito standard
        let statusByte = UInt8(langBytes.count & 0x3F) // Specifica NDEF Text Status Byte
        
        let payloadLength = 1 + langBytes.count + textBytes.count
        
        var record: [UInt8] = []
        record.append(header)
        record.append(0x01) // Type Length ("T")
        record.append(UInt8(payloadLength & 0xFF))
        record.append(0x54) // Type: "T" (RTD_TEXT in ASCII)
        record.append(statusByte)
        record.append(contentsOf: langBytes)
        record.append(contentsOf: textBytes)
        return record
    }
}

File: Python/SmartCardRWPN532/NdefEncoder.py
#**************************************************************************************#
# Nome Modulo: NdefEncoder.py
# Descrizione: Helper specifico per la compilazione e la formattazione binaria TLV NDEF.
#**************************************************************************************#

class NdefEncoder:
    @staticmethod
    def compile_records(records_data):
        """
        Prende una lista di dizionari del tipo: [{'type': 'T' o 'U', 'value': 'stringa'}]
        Genera l'array di byte completo racchiuso nel contenitore TLV (0x03 ... 0xFE).
        """
        ndef_payload = []
        total_records = len(records_data)
        
        if total_records == 0:
            return []

        for idx, rec in enumerate(records_data):
            rec_type = rec['type']   # "T" (Testo) o "U" (URI/URL)
            value_str = rec['value']
            
            # 1. COMPOSIZIONE PAYLOAD SPECIFICO DEL RECORD
            record_payload = []
            if rec_type == "U":
                # Semplificazione: Assumiamo il prefisso standard "https://www." (0x02)
                # Rimuoviamo il prefisso dalla stringa se l'utente lo ha inserito manualmente
                clean_value = value_str.replace("https://www.", "").replace("http://www.", "")
                record_payload.append(0x02) # Prefisso binario standard
                record_payload.extend([ord(c) for c in clean_value])
            elif rec_type == "T":
                # Standard Text: 1 byte status (lunghezza lingua = 2 per "en") + codice lingua + testo
                record_payload.append(0x02) # 0x02 significa: UTF-8, lunghezza codice lingua = 2
                record_payload.extend([ord('e'), ord('n')]) # Codice lingua fisso "en"
                record_payload.extend([ord(c) for c in value_str.encode('utf-8', errors='replace').decode('utf-8')])

            # 2. CALCOLO DEI BIT DI CONTROLLO DELL'HEADER NDEF (Flags Byte)
            mb = 1 if idx == 0 else 0                   # Message Begin (1 solo per il primo record)
            me = 1 if idx == (total_records - 1) else 0 # Message End (1 solo per l'ultimo record)
            cf = 0                                      # Chunk Flag (No chunking)
            sr = 1 if len(record_payload) <= 255 else 0 # Short Record (1 se payload <= 255 byte)
            il = 0                                      # ID Length Present (No ID)
            tnf = 0x01                                  # Type Name Format: NFC Forum Well-Known

            flags_byte = (mb << 7) | (me << 6) | (cf << 5) | (sr << 4) | (il << 3) | tnf
            
            # Assemblaggio intestazione record
            ndef_payload.append(flags_byte)
            ndef_payload.append(len(rec_type)) # Lunghezza del tipo (sempre 1 per "T" o "U")
            
            if sr == 1:
                ndef_payload.append(len(record_payload)) # 1 byte lunghezza payload
            else:
                p_len = len(record_payload)
                ndef_payload.extend([
                    (p_len >> 24) & 0xFF,
                    (p_len >> 16) & 0xFF,
                    (p_len >> 8) & 0xFF,
                    p_len & 0xFF
                ])
                
            ndef_payload.append(ord(rec_type)) # Inserimento del tipo ("T" o "U")
            ndef_payload.extend(record_payload) # Inserimento del payload reale

        # 3. INCAPSULAMENTO NEL BLOCCO CONTENITORE TLV (Type 0x03)
        tlv_frame = []
        tlv_frame.append(0x03) # Type: NDEF Message
        
        ndef_len = len(ndef_payload)
        if ndef_len <= 254:
            tlv_frame.append(ndef_len)
        else:
            tlv_frame.append(0xFF) # Indicatore di lunghezza estesa a 3 byte
            tlv_frame.append((ndef_len >> 8) & 0xFF)
            tlv_frame.append(ndef_len & 0xFF)
            
        tlv_frame.extend(ndef_payload)
        tlv_frame.append(0xFE) # Terminator TLV obbligatorio NFC Forum

        # 4. ALLINEAMENTO AL PADDING A PAGINE DA 4 BYTE
        while len(tlv_frame) % 4 != 0:
            tlv_frame.append(0x00) # Riempiamo l'ultima pagina incompleta con byte nulli

        return tlv_frame
